hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java

```diff
@@ -0,0 +1,674 @@
+      @NotNull
+      @Override
+      protected XceiverClientFactory createXceiverClientFactory(
+          List<X509Certificate> x509Certificates) throws IOException {
+        return mockDataNodeClientFactory();
+      }
+    };
+
+    objectStore = new ObjectStore(conf, rpcClient);
+
+    createVolume(VOLUME_NAME);
+    createBucket(VOLUME_NAME, BUCKET_NAME, false);
+    createBucket(VOLUME_NAME, VERSIONED_BUCKET_NAME, true);
+  }
+
+  @AfterAll
+  public static void cleanup() throws Exception {
+    om.stop();
+    FileUtils.deleteDirectory(dir);
+  }
+
+  private static XceiverClientManager mockDataNodeClientFactory()
+      throws IOException {
+    mockDn1Protocol = spy(new XceiverClientGrpc(createPipeline(DN1), conf));
+    mockDn2Protocol = spy(new XceiverClientGrpc(createPipeline(DN2), conf));
+    XceiverClientManager manager = mock(XceiverClientManager.class);
+    when(manager.acquireClient(argThat(matchPipeline(DN1))))
+        .thenReturn(mockDn1Protocol);
+    when(manager.acquireClientForReadData(argThat(matchPipeline(DN1))))
+        .thenReturn(mockDn1Protocol);
+
+    when(manager.acquireClient(argThat(matchPipeline(DN2))))
+        .thenReturn(mockDn2Protocol);
+    when(manager.acquireClientForReadData(argThat(matchPipeline(DN2))))
+        .thenReturn(mockDn2Protocol);
+    return manager;
+  }
+
+  private static ArgumentMatcher<Pipeline> matchPipeline(DatanodeDetails dn) {
+    return argument -> argument != null
+        && argument.getNodes().get(0).getUuid().equals(dn.getUuid());
+  }
+
+  private static void createBucket(String volumeName, String bucketName,
+                                   boolean isVersionEnabled)
+      throws IOException {
+    OmBucketInfo bucketInfo = OmBucketInfo.newBuilder()
+        .setVolumeName(volumeName)
+        .setBucketName(bucketName)
+        .setIsVersionEnabled(isVersionEnabled)
+        .build();
+
+    OMRequestTestUtils.addBucketToOM(metadataManager, bucketInfo);
+  }
+
+  private static void createVolume(String volumeName) throws IOException {
+    OmVolumeArgs volumeArgs = OmVolumeArgs.newBuilder()
+        .setVolume(volumeName)
+        .setAdminName("bilbo")
+        .setOwnerName("bilbo")
+        .build();
+    OMRequestTestUtils.addVolumeToOM(metadataManager, volumeArgs);
+  }
+
+  @BeforeEach
+  @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
+  public void beforeEach() {
+    testContainerId++;
+    Mockito.reset(mockScmBlockLocationProtocol, mockScmContainerClient,
+        mockDn1Protocol, mockDn2Protocol);
+    when(mockDn1Protocol.getPipeline()).thenReturn(createPipeline(DN1));
+    when(mockDn2Protocol.getPipeline()).thenReturn(createPipeline(DN2));
+  }
+
+  /**
+   * Verify that in a happy case, container location is cached and reused
+   * in OM.
+   */
+  @Test
+  public void containerCachedInHappyCase() throws Exception {
+    byte[] data = "Test content".getBytes(UTF_8);
+
+    mockScmAllocation(testContainerId, 1L);
```

Comment:
This method name could be clearer.  Instead of mockScmAllocation(), how about mockAllocateBlockOnDN1()  ?